### ISSUE TITLE
p:resolve-uri and xml:base

### ISSUE THREAD

**Issue by apb2006** [ndw/xmlcalabash1#33]
There seems to be a problem with the handling of xml:base in XML Calabash version 1.0.2 (for Saxon 9.4.x),  
The pipeline below displays:

```
 <c:request xmlns:c="http://www.w3.org/ns/xproc-step" method="GET" detailed="true" 
      href="file:/C:/saxon/xproc.20/?doc=123"/>
```

In Calabash version 0.9.21 it displayed what I expected..

```
<c:request xmlns:c="http://www.w3.org/ns/xproc-step" method="GET" detailed="true" 
       href="http://xmlcalabash.com/?doc=123"/>
```

Pipeline

```
    <?xml version="1.0"?>
    <p:declare-step name="pipeline" version="1.0"
      xmlns:p="http://www.w3.org/ns/xproc" 
      xmlns:c="http://www.w3.org/ns/xproc-step">
      <p:output port="result"/>
        <p:variable name="item" select="123"/>
         <p:string-replace match="/c:request/@href">
          <p:input port="source">
            <p:inline>
              <c:request method="GET" detailed="true" href="@@HERE@@" />
            </p:inline>
          </p:input>
          <p:with-option name="replace"
            select="concat('&quot;'
                           ,p:resolve-uri(/urls/req[@type='doc']/@url)
                           ,$item
                           ,'&quot;')">
            <p:inline>
              <urls xml:base="http://xmlcalabash.com/">
                <req type="doc" url="?doc=" />
                <req type="compID" url="?compID=" />
              </urls>
            </p:inline>
          </p:with-option>
        </p:string-replace>
    </p:declare-step>
```

**Comment by ndw** [comment]
Unfortunately, I think the bug was in 0.9.21. The problem is that the p:resolve-uri() function is resolved against the base URI of the _document_ that is the context node. The base URI of the p:inline _document_ is the same as the URI of the p:inline element. The xml:base attribute doesn't change the base URI until you get to the root _element_.

There are two ways to fix your pipeline:

First, if you know the base URI statically, you can use it in the p:resolve-uri call:

  p:resolve-uri(/urls/req[@type='doc']/@url, 'http://xmlcalabash.com/')

Second, you can set the base URI on the p:inline:

```
   ...
    <p:inline xml:base="http://xmlcalabash.com/">
      <urls>
      ...
```

That will change the base URI of the document node.

Does that help?

**Comment by apb2006** [comment]
Ok, thanks I had already gone for the 1st option, but I not sure I really get it. Time to read those specs again :-)
This gives my desired result in version 1.0.2

```
  select="concat('&quot;'
                       ,p:resolve-uri(/urls/req[@type='doc']/@url
                                    ,p:base-uri(/urls/req[@type='doc']/@url))
                   ,$item
                   ,'&quot;')">
```

My working intuition has been that these are always the same

```
  resolve-uri(node)<-->resolve-uri(node,base-uri(node))
```

**Comment by apb2006** [comment]
With Calumet 1.0.12 the results seem to agree with  the latest Calabash for the 1 argument form of p:resolve-uri but it calculates  p:base-uri(/urls/req[@type='doc']/@url) as the empty string, giving a result for my second example of

```
 <c:request detailed="true" href="?doc=123" method="GET" xmlns:c="http://www.w3.org/ns/xproc-step"/>
```

**Comment by ndw** [comment]
No, resolve-uri(node) resolves, uh, string(node) against the base URI from the static context.
Whereas resolve-uri(node,base-uri(node)) resolves string(node) against the base URI of node.

I think the example that gives an empty string is a bug in Calumet, but I could be mistaken. I'll point Vojtech here.

(Hi Vojtech :-)

**Comment by vojtechtoman** [comment]
As for Calumet 1.0.12 returning an empty string as the result of p:base-uri(/urls/req[@type='doc']/@url), that is a bug that I have apparently fixed in 1.0.13 (the last released version).

As for the original issue (p:resolve-uri(/urls/req[@type='doc']/@url) returning the base URI of the p:inline element and not the base URI of the "urls" element), I think this is the correct behavior (although maybe confusing for the users) and Norm is right.

**Comment by apb2006** [comment]
Thanks both for the explanation.It begins to make sense to me, intuition updated.
